### DISCUSSION TITLE
For `tuple.map`, add `Tuple.Union[this.type]` as upper bound for polymorphic function `f`

### DISCUSSION THREAD

**Discussion by felher** [scala/scala3#21940] [Feature Requests]
Currently, the signature of `map` on tuples is:
```scala
inline def map[F[_]](f: [t] => t => F[t]): Map[this.type, F]
```
but it seems that it could actually be
```scala
inline def map[F[_]](f: [t <: Tuple.Union[this.type]] => t => F[t]): Map[this.type, F]
```

this would make it possible, for example, to `map` over a tuple of `Int`s and increment each field by one.

**Comment by SethTisue** [comment]
someone on Discord pointed out that @odersky previously suggested changing the result type of `map` at https://github.com/scala/scala3/issues/15992#issuecomment-1243700382 — perhaps that's orthogonal to this? or perhaps the two changes should be considered together?

**Comment by dbear496** [comment]
I think this change would effectively require implementing a `Product.Union` type and reimplementing `Product.productIterator` and `Product.productElement` to return `Iterator[Product.Union[this.type]]` and `Product.Union[this.type]` respectively. Alternatively, `Tuple` could override those methods to change the return type in a similar way. This is because `Tuple.map` uses the `productIterator` method to iterate over the elements of the tuple and feed them one-by-one to `f`.
